File: solutions/python/two-bucket/1/two_bucket.py
def gcdx(a, b):
    """ extended gcd of a, b 
        returns gcd and coefficents u,v 
        s.t.
        gcd = u*a + v*b    
    """
    if b==0:
        gcd = a
        u = 1
        v = 0
    else:
        gcd, u1, u = gcdx(b, a%b)        
        v = u1 - u * (a//b)
    return gcd, u, v

def measure_one(bucket_one, bucket_two, target_volume):
    """
    solves the two-bucket problem
    for the case that bucket one is the first to fill 

    input: size of bucket one,
        size of bucket two,
        desired number of liters to reach

    output:
        total number of steps,
        which bucket holds the desired number of liters,
        how many liters are left in the other bucket     
    """
    if target_volume == bucket_one:
        return (1,"one", 0)
    elif target_volume == bucket_two:
        return  (2,"two", bucket_one)

    d,u,v = gcdx(bucket_one, bucket_two)
    if target_volume%d!=0:
        raise ValueError('Not possible.')
    u *= target_volume//d
    v *= target_volume//d

    if u<0:
        t = -u*d//bucket_two + 1
        u += (bucket_two//d)*t
        v -= (bucket_one//d)*t
    if bucket_one>=target_volume:
        return (abs(2*(u-v)-2), "one", bucket_two)
    else:
        return  (abs(2*(u-v)), "two", 0)
        

def measure(bucket_one, bucket_two, goal, start_bucket):
    """
    solves the two-bucket problem
    
    input: size of bucket one,
        size of bucket two,
        desired number of liters to reach,
        which bucket is first to fill

    output:
        total number of steps,
        which bucket holds the desired number of liters,
        how many liters are left in the other bucket    
    """
    if goal == 0:
        return (0,"one", 0)
    elif goal<0 or goal>max(bucket_one, bucket_two):
        raise ValueError('Not possible.')
    elif start_bucket == "one":
        return measure_one(bucket_one, bucket_two, goal)
    else:
        moves, target, other_vol = measure_one(bucket_two, bucket_one, goal)
        if target == "one":
            target = "two"
        elif target == "two":
            target = "one"
        return (moves, target, other_vol)

File: solutions/free-pascal/two-bucket/2/TwoBucket.pas
unit TwoBucket;

{$mode ObjFPC}{$H+}

interface

type
  BucketId = (One, Two);

  TResult = record
    moves: UInt64;
    goalBucket: BucketId;
    otherBucket: UInt64;
  end;

function measure(bucketOne, bucketTwo, goal : UInt64; startBucket: BucketId) : TResult;

implementation

uses SysUtils, Math;

(* Use extended gcd to determine Bezout coefficients u,v s.t. 
d = u*a + v*b
*)

type 
TGcdX = record
  d : UInt64;
  u : Int64;
  v : Int64;
end;

function gcdx(a,b : UInt64) : TGcdX;
var 
  u1,v1 : Int64;
begin
  if b = 0 then
   with result do
   begin
     d := a;
     u := 1; 
     v := 0;
   end
  else
    begin
      result := gcdx(b, (a mod b));
      v1 :=  result.v;
      u1 :=  result.u;
      result.u:= v1;
      result.v:= u1 - v1 * (a div b);
    end;    
end;

(* Solve the two bucket problem for the case that bucket One is the first to fill *)

function measureOne(bOne, bTwo, goal : UInt64) : TResult;
var gcd: TGcdx;
    t : Int64;
begin
  if goal = bOne then
    with result do
      begin
        moves := 1;
        goalBucket:=One;
        otherBucket := 0;
      end
  else if goal = bTwo then
    with result do
      begin
        moves := 2;
        goalBucket := Two;
        otherBucket := bOne;
      end 
  else
  begin
    gcd := gcdx(bOne, bTwo);
    if (goal mod gcd.d)<>0 then
      raise Exception.create('impossible');
    (* rescale coefficients s.t. 
    u*b1 + v*b2 = goal *)
    gcd.u := gcd.u * (goal div gcd.d);
    gcd.v := gcd.v * (goal div gcd.d);
    if gcd.u < 0 then
    begin
      t:= ((-gcd.u *gcd.d) div bTwo)+1;
      gcd.u := gcd.u + (bTwo div gcd.d)*t;
      gcd.v := gcd.v - (bOne div gcd.d)*t;
    end;
    if bOne>=goal then 
    (* If target volume fits in bucket One,
    we can elimnate the last last "empty two"
    and "pour into two" op.*)
      with result do
      begin
        moves := abs(2*(gcd.u - gcd.v) - 2);
        goalBucket := One;
        otherBucket := bTwo;
      end
    else
      with result do
      begin 
        moves := abs(2*(gcd.u - gcd.v));
        goalBucket := Two;
        otherBucket := 0;  
      end;
    end;
end;

function measure(bucketOne, bucketTwo, goal : UInt64; startBucket: BucketId) : TResult;
begin
  if goal = 0 then
    with result do
    begin
      moves := 0;
      goalBucket := startBucket;
      otherBucket := 0;
    end;
  if goal>max(bucketOne, bucketTwo) then
    raise Exception.create('impossible');
    
  if startBucket = One then 
    result:=measureOne(bucketOne,bucketTwo,goal)
  else
  (*Start bucket is Two: Switch the buckets and switch back id of start bucket*)
    begin
      result:=measureOne(bucketTwo, bucketOne, goal);
      if result.goalBucket=One then
        result.goalBucket:=Two
      else
        result.goalBucket:=One;
    end;
end;

end.
